### PR TITLE
Support links to be returned from queries

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Adds support to return `Link`s referred to from entities, either by starting from an entity query or from a link query

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1202882583267702/1202986163519979/f) _(internal)_

## 🔍 What does this change?

- Unify `EntityQuery` and `LinkQuery` into `KnowledgeQuery`
- Support links to be returned from entities
- Support linked entities returned from Link queries

## 📜 Does this require a change to the docs?

The OpenAPI client was updated

## ⚠️ Known issues

The return structure is not necessary optimal, we might want to [move to a vertex/edge set](https://app.asana.com/0/0/1202996107497595/f) (_internal_)

## 🛡 What tests cover this?

Test the query with different setups and parameters

## 📹 Demo

```http
POST http://127.0.0.1:4000/entities/query
Content-Type: application/json

{
  "query": {
    "eq": [
      {
        "path": [
          "version"
        ]
      },
      {
        "literal": "latest"
      }
    ]
  },
  "dataTypeQueryDepth": 0,
  "propertyTypeQueryDepth": 0,
  "linkTypeQueryDepth": 1,
  "entityTypeQueryDepth": 0,
  "linkTargetEntityQueryDepth": 10,
  "linkQueryDepth": 10
}
```

```json
[
  {
    "entity": {
      "inner": {
        "http://localhost:3000/@alice/types/property-type/name/": "Alice Allison"
      },
      "identifier": {
        "entityId": "0335be0b-26a1-475f-81de-aabcb3bf0676",
        "version": "2022-09-16T10:21:38.140412Z",
        "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
      },
      "typeVersionedUri": "http://localhost:3000/@alice/types/entity-type/person/v/2"
    },
    "referencedDataTypes": [],
    "referencedPropertyTypes": [],
    "referencedLinkTypes": [
      {
        "inner": {
          "$id": "http://localhost:3000/@alice/types/link-type/friend-of/v/1",
          "description": "Someone who has a shared bond of mutual affection",
          "kind": "linkType",
          "pluralTitle": "Friends Of",
          "title": "Friend Of"
        },
        "identifier": {
          "uri": "http://localhost:3000/@alice/types/link-type/friend-of/v/1",
          "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
        }
      }
    ],
    "referencedEntityTypes": [],
    "linkedEntities": [
      {
        "inner": {
          "http://localhost:3000/@alice/types/property-type/name/": "Bob"
        },
        "identifier": {
          "entityId": "c0950d7e-68ee-41c8-a9bd-e1e2dddd5a2d",
          "version": "2022-09-16T10:21:38.254749Z",
          "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
        },
        "typeVersionedUri": "http://localhost:3000/@alice/types/entity-type/person/v/1"
      }
    ],
    "links": [
      {
        "sourceEntityId": "c0950d7e-68ee-41c8-a9bd-e1e2dddd5a2d",
        "targetEntityId": "0335be0b-26a1-475f-81de-aabcb3bf0676",
        "linkTypeUri": "http://localhost:3000/@alice/types/link-type/friend-of/v/1"
      },
      {
        "sourceEntityId": "0335be0b-26a1-475f-81de-aabcb3bf0676",
        "targetEntityId": "c0950d7e-68ee-41c8-a9bd-e1e2dddd5a2d",
        "linkTypeUri": "http://localhost:3000/@alice/types/link-type/friend-of/v/1"
      }
    ]
  },
  {
    "entity": {
      "inner": {
        "http://localhost:3000/@alice/types/property-type/name/": "Bob"
      },
      "identifier": {
        "entityId": "c0950d7e-68ee-41c8-a9bd-e1e2dddd5a2d",
        "version": "2022-09-16T10:21:38.254749Z",
        "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
      },
      "typeVersionedUri": "http://localhost:3000/@alice/types/entity-type/person/v/1"
    },
    "referencedDataTypes": [],
    "referencedPropertyTypes": [],
    "referencedLinkTypes": [
      {
        "inner": {
          "$id": "http://localhost:3000/@alice/types/link-type/friend-of/v/1",
          "description": "Someone who has a shared bond of mutual affection",
          "kind": "linkType",
          "pluralTitle": "Friends Of",
          "title": "Friend Of"
        },
        "identifier": {
          "uri": "http://localhost:3000/@alice/types/link-type/friend-of/v/1",
          "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
        }
      }
    ],
    "referencedEntityTypes": [],
    "linkedEntities": [
      {
        "inner": {
          "http://localhost:3000/@alice/types/property-type/name/": "Alice Allison"
        },
        "identifier": {
          "entityId": "0335be0b-26a1-475f-81de-aabcb3bf0676",
          "version": "2022-09-16T10:21:38.140412Z",
          "ownedById": "74014335-79d2-4a66-8533-5a9165538497"
        },
        "typeVersionedUri": "http://localhost:3000/@alice/types/entity-type/person/v/2"
      }
    ],
    "links": [
      {
        "sourceEntityId": "0335be0b-26a1-475f-81de-aabcb3bf0676",
        "targetEntityId": "c0950d7e-68ee-41c8-a9bd-e1e2dddd5a2d",
        "linkTypeUri": "http://localhost:3000/@alice/types/link-type/friend-of/v/1"
      },
      {
        "sourceEntityId": "c0950d7e-68ee-41c8-a9bd-e1e2dddd5a2d",
        "targetEntityId": "0335be0b-26a1-475f-81de-aabcb3bf0676",
        "linkTypeUri": "http://localhost:3000/@alice/types/link-type/friend-of/v/1"
      }
    ]
  }
]
```

### CODE REVIEW

File: packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs [hunk 1]
```diff
@@ -19,12 +19,12 @@ use crate::{
...
 };
 
-/// Query to read [`Entities`], which satisfy the [`Expression`].
+/// Query to read [`Entities`] or [`Link`]s, which satisfy the [`Expression`].
 ///
 /// [`Entities`]: Entity
 #[derive(Debug, Deserialize, Component)]
 #[serde(deny_unknown_fields, rename_all = "camelCase")]
-pub struct EntityQuery {
+pub struct KnowledgeQuery {
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs:27]: Maybe call it `KnowledgeGraphQuery`?

File: packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs [hunk 2]
```diff
@@ -36,7 +36,9 @@ pub struct EntityQuery {
     #[component(value_type = number)]
     pub entity_type_query_depth: QueryDepth,
     #[component(value_type = number)]
-    pub linked_entity_query_depth: QueryDepth,
+    pub link_target_entity_query_depth: QueryDepth,
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs:39]: I missed this before but do the `QueryDepth` docs still only talk about ontology elements? We might want to either make a separate alias or fix the docs

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/mod.rs [hunk 3]
```diff
@@ -1,16 +1,90 @@
...
+        postgres::{
+            knowledge::KnowledgeDependencyContext, ontology::LinkTypeDependencyContext,
+            DependencyMap, DependencySet,
+        },
+        AsClient, InsertionError, LinkStore, PostgresStore, QueryError,
+    },
 };
 
+impl<C: AsClient> PostgresStore<C> {
+    /// Internal method to read a [`PersistedDataType`] into a [`DependencyMap`].
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/mod.rs]: PersistedDataType?

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/mod.rs]: I see you've fixed this

File: packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs (hunk 2 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs:39]: I think I will add a new alias, in the case we change the type, it's likely that it's only changed for either the ontology or the knowledge

File: packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs (hunk 1 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs]: Applied in ee984bdab2943a042988130478e2fedce93e30d4

File: packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs [hunk 4]
```diff
@@ -13,30 +13,59 @@ pub use self::{
...
+        PersistedPropertyType,
     },
     store::query::Expression,
 };
 
-/// Query to read [`Entities`], which satisfy the [`Expression`].
+/// Distance to explore when querying a rooted subgraph on entities and links.
+///
+/// Entities may link to other entities through links. The depths provided alongside a query specify
+/// how many steps to explore along a chain of references _of a certain kind of type_. Meaning, any
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs:25]: These aren't chains of references, or types

File: packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs (hunk 2 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs]: Added in https://github.com/hashintel/hash/pull/1074/commits/c92b8910eb27c5dcedb1fad949ef8fbdc4f2b0eb, also adjusted the documentation to match the behavior. Note: it's the same behavior, but entities and links are chained differently from ontology types. I struggled a bit with the documentation, feel free to adjust it 🙂

File: packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs [hunk 5]
```diff
@@ -13,30 +13,59 @@ pub use self::{
...
+/// chain of entities and links will be resolved up to the given depth. These can be composed with
+/// [`OntologyQueryDepth`] to explore ontology types.
+///
+/// A depth of `0` means that links are explored for that specific kind of type.
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs:29]: Don't really know what the last bit mean

File: packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs [hunk 6]
```diff
@@ -13,30 +13,59 @@ pub use self::{
...
+/// - `Entity2` links to `Entity3` by `Link2`
+///
+/// If a query on `Entity1` is made with the following depths:
+/// - `link_query_depth: 2`
+/// - `link_target_entity_query_depth: 1`
+///
+/// the query will resolve up to two links, but only a single entity:
+/// - `linkedEntities`: \[`Entity2`]
+/// - `links`: \[`Link1`, `Link2`]
+pub type KnowledgeQueryDepth = u8;
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs:47]: Maybe we should call it KnowledgeGraphQueryDepth?

I don't feel strongly about this one

Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/knowledge/mod.rs]: It's good I guess, then it's consistent with `KnowledgeGraphQuery`: 695278dcca9caccdd59b309c856baaa1feda2d85

**Review by Alfred-Mountfield — APPROVED** [review]
🚀

**Review by CiaranMn — APPROVED** [review]

**Pull request merged**